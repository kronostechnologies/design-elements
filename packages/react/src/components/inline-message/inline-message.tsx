import React, { ComponentType, ReactElement, ReactNode, useMemo } from 'react';
import { useTranslation } from 'react-i18next';
import styled from 'styled-components';

import { i18n } from '../../i18n/i18n';
import { useDeviceContext } from '../device-context-provider/device-context-provider';
import { Icon, IconName } from '../icon/icon';
import { Theme } from '../theme-wrapper/theme-wrapper';

type MessageType = 'info' | 'success' | 'alert' | 'error';

const abstractContainer = (bgColor: string, color?: keyof Theme['notifications']) => styled.div<{isMobile: boolean}>`
    background-color: ${bgColor};
    border: 1px solid ${props => color ? props.theme.notifications[color] : props.theme.main['primary-3']};
    box-sizing: border-box;
    display: flex;
    padding: ${props => props.isMobile ? 'var(--spacing-3x) var(--spacing-2x)' : 'var(--spacing-2x)'};
    width: 100%;

    svg {
        color: ${props => color ? props.theme.notifications[color] : props.theme.main['primary-3']};
        flex: 0 0 auto;
    }
`;

const InfoContainer = abstractContainer('#f5fdff');
const SuccessContainer = abstractContainer('#f7faf4', 'success-1.1');
const AlertContainer = abstractContainer('#fffce9', 'alert-3.1');
const ErrorContainer = abstractContainer('#fdf6f7', 'error-2.1');

const TextWrapper = styled.div<{isMobile: boolean}>`
    box-sizing: border-box;
    padding-left: var(--spacing-2x);

    p {
        font-size: ${props => props.isMobile ? '1rem' : '0.875rem'};
        line-height: 24px;
        margin: ${props => props.isMobile ? 'var(--spacing-2x)' : 'var(--spacing-1x)'} 0 0 0;
    }
`;

const Heading = styled.span<{isMobile: boolean}>`
    font-size: ${props => props.isMobile ? '1.125rem' : '1rem'};
    font-weight: var(--font-bold);
`;

interface MessageTypeProps {
    container: ComponentType<{ isMobile: boolean; }>;
    iconName: IconName;
    title: string;
}

const handleType = (type: MessageType): MessageTypeProps => {
    switch (type) {
        case 'info':
            return {
                container: InfoContainer,
                iconName: 'star',
                title: 'Tips',
            };
        case 'success':
            return {
                container: SuccessContainer,
                iconName: 'check',
                title: 'Success',
            };
        case 'alert':
            return {
                container: AlertContainer,
                iconName: 'alertTriangle',
                title: 'Alert',
            };
        case 'error':
            return {
                container: ErrorContainer,
                iconName: 'alertOctagon',
                title: 'Error',
            };
    }
};

interface InlineMessageProps {
    /**
     * Sets text message
     **/
    children: ReactNode;
    /**
     * Sets custom message title
     */
    title?: string;
    /**
     * Sets message type
     */
    type: MessageType;
}

export function InlineMessage({ children, title, type }: InlineMessageProps): ReactElement {
    const { t } = useTranslation('inline-message');
    const { isMobile } = useDeviceContext();
    const messageType: MessageTypeProps = useMemo(() => handleType(type), [type]);
    const Container = messageType.container;

    return (
        <Container isMobile={isMobile} aria-live={type === 'alert' || type === 'error' ? 'assertive' : 'polite'}>
            <Icon name={messageType.iconName} size={isMobile ? '20' : '16'}/>
            <TextWrapper isMobile={isMobile}>
                <Heading isMobile={isMobile}>{title ? title : t(messageType.title)}</Heading>
                <p>{children}</p>
            </TextWrapper>
        </Container>
    );
}

const Translation = {
    en: {
        Tips: 'Tips',
        Success: 'Success',
        Alert: 'Alert',
        Error: 'Error',
    },
    fr: {
        Tips: 'Astuce',
        Success: 'Succ√®s',
        Alert: 'Alerte',
        Error: 'Erreur',
    },
};

i18n.addResources('en', 'inline-message', Translation.en);
i18n.addResources('fr', 'inline-message', Translation.fr);
