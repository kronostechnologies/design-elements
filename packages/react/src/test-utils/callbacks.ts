export function doNothing(): void {
    // do nothing
}
