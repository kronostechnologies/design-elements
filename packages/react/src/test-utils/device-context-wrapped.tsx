import React, { ReactElement } from 'react';
import { DeviceContextProvider, DeviceType } from '../components/device-context-provider/device-context-provider';

export function DeviceContextWrapped(children: ReactElement, device?: DeviceType): ReactElement {
    return (
        <DeviceContextProvider staticDevice={device}>
            {children}
        </DeviceContextProvider>
    );
}
