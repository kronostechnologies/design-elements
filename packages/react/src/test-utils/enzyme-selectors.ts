import { ReactWrapper, ShallowWrapper } from 'enzyme';

enum ModifierEnum {
    StartsWith = '^',
    EndsWith = '$',
    ContainsOptionalHyphen = '|',
    Contains = '*',
    ContainsWord = '~',
    None = '',
}
type SelectorModifier = `${ModifierEnum}`;

function filterHtmlNodesOnly<W extends ReactWrapper | ShallowWrapper>(
    nodes: W,
): W {
    return nodes
        .hostNodes()
        .filterWhere((x) => x.name() !== null && !x.name().startsWith('styled.')) as W;
}

export function findByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node over the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    const modifier = options?.modifier ?? '';

    let nodes = wrapper.find(`[data-testid${modifier}="${testId}"]`) as W;

    if (options?.nodeName) {
        nodes = nodes.filterWhere((x) => x.name() === options.nodeName) as W;
    }
    if (options?.htmlNodesOnly) {
        nodes = filterHtmlNodesOnly(nodes);
    }

    return nodes;
}

export function getByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node instead of the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    return findByTestId(wrapper, testId, options).at(0) as W;
}
