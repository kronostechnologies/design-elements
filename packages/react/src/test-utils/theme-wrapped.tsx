import React, { ComponentProps, ComponentType, ReactElement } from 'react';
import { equisoftTheme, ThemeWrapper } from '..';

type ThemeProps = Pick<ComponentProps<typeof ThemeWrapper>, 'children'>;

export function themeProvider(): ComponentType {
    return  ThemeProvider as ComponentType;
}

export function ThemeProvider({ children }: ThemeProps): ReactElement {
    return (
        <ThemeWrapper theme={equisoftTheme}>
            {children}
        </ThemeWrapper>
    );
}

export function ThemeWrapped(children: ReactElement): ReactElement {
    return (
        <ThemeWrapper theme={equisoftTheme}>
            {children}
        </ThemeWrapper>
    );
}
