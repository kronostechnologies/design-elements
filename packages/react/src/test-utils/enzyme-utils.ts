import { ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';

export async function waitForComponentToPaint<P = {}>(
    wrapper: ReactWrapper<P>,
    amount = 0,
): Promise<void> {
    await act(async () => {
        await new Promise((resolve) => setTimeout(resolve, amount));
        wrapper.update();
    });
}

export function expectFocusToBeOn(element: ReactWrapper): void {
    expect(document.activeElement).toBe(element.getDOMNode());
}
