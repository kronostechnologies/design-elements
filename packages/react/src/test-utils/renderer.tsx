import { ThemeWrapped } from '@design-elements/test-utils/theme-wrapped';
import { mount, ReactWrapper, render } from 'enzyme';
import React, { Component, ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { DeviceContextProvider } from '../components/device-context-provider/device-context-provider';

export function mountWithProviders<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
): ReactWrapper<P, S, C> {
    return mount(component, {
        wrappingComponent: AllProviders,
    });
}

export function renderWithProviders(
    component: ReactElement,
): Cheerio {
    return render(<AllProviders>{component}</AllProviders>);
}

function AllProviders({ children }: { children: ReactElement }): ReactElement {
    return (
        <MemoryRouter>
            <DeviceContextProvider>
                {ThemeWrapped(children)}
            </DeviceContextProvider>
        </MemoryRouter>
    );
}
