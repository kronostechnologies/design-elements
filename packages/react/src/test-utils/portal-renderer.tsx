import { render, RenderResult } from '@testing-library/react';
import React, { ReactElement } from 'react';
import { DeviceType } from '../components/device-context-provider/device-context-provider';
import { AllProviders } from './renderer';

export function renderPortalWithProviders(
    component: ReactElement,
    device?: DeviceType,
): RenderResult {
    return render(<AllProviders device={device}>{component}</AllProviders>);
}
