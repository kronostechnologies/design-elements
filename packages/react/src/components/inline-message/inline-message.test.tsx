import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeWrapped } from '../../test-utils/theme-wrapped';
import { InlineMessage } from './inline-message';

describe('Inline Message', () => {
    test('Matches snapshot (desktop)', () => {
        const tree = renderer.create(
            ThemeWrapped(
                <InlineMessage type="info">
                    Test
                </InlineMessage>,
            ),
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    test('Matches snapshot (mobile)', () => {
        const tree = renderer.create(
            ThemeWrapped(
                <InlineMessage type="error" device="mobile">
                    Test
                </InlineMessage>,
            ),
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
