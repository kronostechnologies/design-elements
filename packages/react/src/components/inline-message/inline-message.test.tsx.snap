// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inline Message Matches snapshot (desktop) 1`] = `
.c0 {
  background-color: #f5fdff;
  border: 1px solid #094C6C;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  max-width: 800px;
  min-height: 88px;
  padding: var(--spacing-2x);
}

.c0 svg {
  color: #094C6C;
}

.c1 {
  box-sizing: border-box;
  max-width: calc(100% - 16px);
  padding-left: 12px;
}

.c1 p {
  font-size: 0.875rem;
  line-height: 24px;
  margin: var(--spacing-1x) 0 0 0;
}

.c2 {
  font-size: 1rem;
  font-weight: var(--font-bold);
}

<div
  className="c0"
  type="info"
>
  <svg
    color="currentColor"
    focusable={false}
    height="16"
    width="16"
  />
  <div
    className="c1"
  >
    <span
      className="c2"
    >
      Tips
    </span>
    <p
      aria-live="polite"
    >
      Test
    </p>
  </div>
</div>
`;

exports[`Inline Message Matches snapshot (mobile) 1`] = `
.c0 {
  background-color: #fdf6f7;
  border: 1px solid #A40C2E;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  max-width: 800px;
  min-height: 88px;
  padding: var(--spacing-3x) var(--spacing-2x);
}

.c0 svg {
  color: #A40C2E;
}

.c1 {
  box-sizing: border-box;
  max-width: calc(100% - 20px);
  padding-left: var(--spacing-2x);
}

.c1 p {
  font-size: 1rem;
  line-height: 24px;
  margin: var(--spacing-2x) 0 0 0;
}

.c2 {
  font-size: 1.125rem;
  font-weight: var(--font-bold);
}

<div
  className="c0"
  type="error"
>
  <svg
    color="currentColor"
    focusable={false}
    height="20"
    width="20"
  />
  <div
    className="c1"
  >
    <span
      className="c2"
    >
      Error
    </span>
    <p
      aria-live="assertive"
    >
      Test
    </p>
  </div>
</div>
`;
